### PR TITLE
feat: Implement webhook object

### PR DESCRIPTION
**Pull request by sillyElvy**
## About

This pull request implements a webhook object. Please test this since I don't have any experience with webhooks.

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [ ] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [ ] I've made this pull request for/as: (check all that apply)
  - [x] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

File: interactions/api/models/webhook.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+        return Message(**res, _client=self._client)
+
+    @property
+    def avatar_url(self) -> Optional[str]:
+        """
+        Returns the URL of the webhook's avatar
+
+        :return: URL of the webhook's avatar.
+        :rtype: str
+        """
```
Comment (Toricane) [inline interactions/api/models/webhook.py]: .
```suggestion
        """
        Returns the URL of the webhook's avatar.

        :return: URL of the webhook's avatar
        :rtype: str
        """
```

**Review by Toricane — CHANGES_REQUESTED** [review]
.

File: interactions/api/models/webhook.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+        )
+
+        return Message(**res, _client=self._client)
+
+    async def delete(self) -> None:
+        """
+        Deletes the webhook.
+        """
+
+        await self._client.delete_webhook(webhook_id=int(self.id), webhook_token=self.token)
```
Comment (Toricane) [inline interactions/api/models/webhook.py]: Add a check to see if self._client exists
```suggestion
    async def delete(self) -> None:
        """
        Deletes the webhook.
        """
        if not self._client:
            raise AttributeError("HTTPClient not found!")

        await self._client.delete_webhook(webhook_id=int(self.id), webhook_token=self.token)
```

File: interactions/api/models/guild.py [hunk 3]
```diff
@@ -1864,6 +1865,14 @@ async def get_all_members(self) -> List[Member]:
 
         return [Member(**_, _client=self._client) for _ in _all_members]
 
+    async def get_webhooks(self) -> List[Webhook]:
+        if not self._client:
+            raise AttributeError("HTTPClient not found!")
+
+        res = await self._client.get_guild_webhooks(int(self.id))
+
+        return [Webhook(**_) for _ in res]
```
Comment (Toricane) [inline interactions/api/models/guild.py]: Add _client?
```suggestion
        return [Webhook(**_, _client=self._client) for _ in res]
```

File: interactions/api/models/channel.py [hunk 4]
```diff
@@ -1106,6 +1107,17 @@ async def get_history(self, limit: int = 100) -> List["Message"]:  # noqa
...
+    async def get_webhooks(self) -> List[Webhook]:
+        """
+        Gets a list of webhooks of that channel
+        """
+
+        if not self._client:
+            raise AttributeError("HTTPClient not found!")
+
+        res = await self._client.get_channel_webhooks(int(self.id))
+        return [Webhook(**_) for _ in res]
```
Comment (Toricane) [inline interactions/api/models/channel.py]: Add _client?
```suggestion
        return [Webhook(**_, _client=self._client) for _ in res]
```

**Review by Toricane — CHANGES_REQUESTED** [review]
Add self._client in various places

**Comment by Toricane** [comment]
How do you test it?
Idk webhooks either lol

**Comment by sillyElvy** [comment]
> How do you test it?
> Idk webhooks either lol

Idk :lol:

**Comment by Damego** [comment]
Error after executing but it works
```py
Task exception was never retrieved
future: <Task finished name='Task-16' coro=<test_command() done, defined at c:\Users\User\OneDrive\ipy-test-bot\main.py:15> exception=TypeError('interactions.api.models.message.Message() argument after ** must be a mapping, not NoneType')>
Traceback (most recent call last):
  File "c:\Users\User\OneDrive\ipy-test-bot\main.py", line 25, in test_command
    await webhook.execute("test message", username="interactions.py webhook", embeds=interactions.Embed(title="konnichiwa"))
  File "c:\Users\User\OneDrive\ipy-test-bot\interactions\api\models\webhook.py", line 274, in execute
    return Message(**res, _client=self._client)
TypeError: interactions.api.models.message.Message() argument after ** must be a mapping, not NoneType
```

**Review by Damego — CHANGES_REQUESTED** [review]
Add webhook import in `models.__init__`

File: interactions/api/models/webhook.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+        payload = {}
+
+        if name is not MISSING:
+            payload["name"] = name
+
+        if avatar is not MISSING:
+            payload["avatar"] = avatar.data
+
+        if not self.token:
```
Comment (Damego) [inline interactions/api/models/webhook.py]: I want change channel but its doesn't allow change channel

```py
webhook = await channel.get_webhooks()[0]
await webhook.modify("name of webhook", 978517248367013948)
await webhook.execute("Hello", username="interactions.py webhook")
 ```

File: interactions/api/models/webhook.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        :type channel_id: int
+        :param name: The name of the webhook.
+        :type name: str
+        :param avatar: The avatar of the Webhook, if any.
+        :type avatar: Optional[Image]
+        :return: The created webhook as object
+        :rtype: Webhook
+        """
+
+        _avatar = avatar if avatar is not MISSING else None
```
Comment (Damego) [inline interactions/api/models/webhook.py]:
```suggestion
        _avatar = avatar.data if avatar is not MISSING else None
```

File: interactions/api/models/webhook.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        :param embeds: embedded ``rich`` content
+        :type embeds: Union[Embed, List[Embed]]
+        :param allowed_mentions: allowed mentions for the message
+        :type allowed_mentions: dict
+        :param components: the components to include with the message
+        :type components: Union[ActionRow, Button, SelectMenu, List[ActionRow], List[Button], List[SelectMenu]]
+        :param files: The files to attach to the message
+        :type files: Union[File, List[File]]
+        :param thread_id: Send a message to a specified thread within a webhook's channel. The thread will automatically be unarchived
+        :type thread_id: int
```
Comment (Damego) [inline interactions/api/models/webhook.py:217]: docs moment: what `execute` should return?

Comment (sillyElvy) [reply interactions/api/models/webhook.py:217]: Optional[Message], i'm fixing it

File: interactions/api/models/webhook.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        :return: The sent message, if present
+        :rtype: Optional[Message
```
Comment (Damego) [inline interactions/api/models/webhook.py]:
```suggestion
        :rtype: Optional[Message]
```

**Review by Damego — APPROVED** [review]
lgtm

**Review by Toricane — APPROVED** [review]
LGTM!

**Pull request merged**